internal/sqlsmith: "index out of range" panic in makeMergeJoinExprwith

internal/sqlsmith.TestGenerateParse [failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=3106126&tab=buildLog) with [artifacts](https://teamcity.cockroachdb.com/viewLog.html?buildId=3106126&tab=artifacts#/) on master @ [2cafdfb574f8f50c0896446b93070aecb1ae87f7](https://github.com/cockroachdb/cockroach/commits/2cafdfb574f8f50c0896446b93070aecb1ae87f7):


```
		AS col_39
LIMIT
	21:::INT8;

STMT: 9
SELECT
	'e84c95bf-7b99-4a96-b8e1-10c063b9893f':::UUID AS col_40, tab_17.tableoid AS col_41, '1991-07-26':::DATE AS col_42
FROM
	defaultdb.public.seed_vec@primary AS tab_17
WHERE
	tab_17._bool;

W210622 15:38:24.870403 11096 sql/sqlliveness/slinstance/slinstance.go:183  [n1] 600  exiting heartbeat loop
W210622 15:38:24.871064 12510 kv/kvserver/intentresolver/intent_resolver.go:786  [-] 601  failed to gc transaction record: could not GC completed transaction anchored at /Table/15/1/669469586835963905: node unavailable; try another peer
W210622 15:38:24.871625 11097 jobs/registry.go:664  [-] 602  canceling all adopted jobs due to stopper quiescing
    panic.go:969: runtime error: index out of range [0] with length 0
        goroutine 10820 [running]:
        runtime/debug.Stack(0xc005bfc908, 0x4c35e60, 0xc00229e360)
        	/usr/local/go/src/runtime/debug/stack.go:24 +0x9f
        github.com/cockroachdb/cockroach/pkg/util/leaktest.AfterTest.func1()
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/leaktest/leaktest.go:110 +0x278
        panic(0x4c35e60, 0xc00229e360)
        	/usr/local/go/src/runtime/panic.go:969 +0x1b9
        github.com/cockroachdb/cockroach/pkg/util/stop.(*Stopper).Stop(0xc0036a2080, 0x5db62c0, 0xc000124010)
        	/go/src/github.com/cockroachdb/cockroach/pkg/util/stop/stopper.go:466 +0x2d3
        panic(0x4c35e60, 0xc00229e360)
        	/usr/local/go/src/runtime/panic.go:969 +0x1b9
        github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.makeMergeJoinExpr.func1(0xc0056c1180, 0xc002682000, 0xc000b45720, 0xc005bfd1d8, 0xc005bfd1a0, 0x0, 0x0, 0x0, 0x0)
        	/go/src/github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:288 +0x8b0
        github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.makeMergeJoinExpr(0xc0056c1180, 0x0, 0x0, 0x0, 0xc005f6e000, 0x12bf1ce405322cd, 0x7ffffffd05bfd2f0, 0xc005bfd2f0, 0x78d405, 0xc002d90660, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:323 +0x26b
        github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.makeTableExpr(0xc0056c1180, 0x0, 0x0, 0x0, 0x6b689500, 0x5599224d20fea199, 0xc005f2cee0, 0x0, 0xc005f38718, 0xc005bfd430, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:144 +0xb0
        github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.(*Smither).makeSelectClause(0xc0056c1180, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:521 +0x186
        github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.(*Smither).makeSelect(0xc0056c1180, 0x0, 0x0, 0x0, 0x0, 0x0, 0x0, 0x300, 0xc005bfd6a8, 0xca, ...)
        	/go/src/github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:726 +0xcf
        github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.makeSelect(0xc0056c1180, 0x5604910, 0xc005a72000, 0xca)
        	/go/src/github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:679 +0x5f
        github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.(*Smither).makeStmt(0xc0056c1180, 0x0, 0x0, 0xca)
        	/go/src/github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/relational.go:20 +0x45
        github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.(*Smither).Generate(0xc0056c1180, 0xc000126008, 0xc005bfde48)
        	/go/src/github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/sqlsmith.go:158 +0x36
        github.com/cockroachdb/cockroach/pkg/internal/sqlsmith.TestGenerateParse(0xc002a83500)
        	/go/src/github.com/cockroachdb/cockroach/pkg/internal/sqlsmith/sqlsmith_test.go:184 +0x732
        testing.tRunner(0xc002a83500, 0x56046f8)
        	/usr/local/go/src/testing/testing.go:1123 +0xef
        created by testing.(*T).Run
        	/usr/local/go/src/testing/testing.go:1168 +0x2b3
--- FAIL: TestGenerateParse (1.44s)
```
<details><summary>Reproduce</summary>
<p>
<p>To reproduce, try:

```bash
make stressrace TESTS=TestGenerateParse PKG=./pkg/internal/sqlsmith TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```
</p>
<p>Parameters in this failure:

- GOFLAGS=-json
</p>
</p>
</details>
<details><summary>Internal log</summary>
<p>

```
mjibson marked as alumn{us/a}; resolving to rafiss instead
```
</p>
</details>

/cc @cockroachdb/sql-queries rafiss
<sub>

[This test on roachdash](https://roachdash.crdb.dev/?filter=status:open%20t:.*TestGenerateParse.*&sort=title+created&display=lastcommented+project) | [Improve this report!](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)
</sub>

--- a/pkg/internal/sqlsmith/BUILD.bazel
+++ b/pkg/internal/sqlsmith/BUILD.bazel
@@ -52,7 +52,6 @@ go_test(
         "//pkg/server",
         "//pkg/sql/parser",
         "//pkg/testutils/serverutils",
-        "//pkg/testutils/skip",
         "//pkg/testutils/sqlutils",
         "//pkg/testutils/testcluster",
         "//pkg/util/leaktest",

--- a/pkg/internal/sqlsmith/schema.go
+++ b/pkg/internal/sqlsmith/schema.go
@@ -438,6 +438,14 @@ func (s *Smither) extractIndexes(
 		if err := rows.Err(); err != nil {
 			return nil, err
 		}
+		// Remove indexes with empty Columns. This is the case for rowid indexes
+		// where the only index column, rowid, is ignored in the SQL statement
+		// above, but the stored columns are not.
+		for name, idx := range indexes {
+			if len(idx.Columns) == 0 {
+				delete(indexes, name)
+			}
+		}
 		ret[*t.TableName] = indexes
 	}
 	return ret, nil

--- a/pkg/internal/sqlsmith/sqlsmith_test.go
+++ b/pkg/internal/sqlsmith/sqlsmith_test.go
@@ -21,7 +21,6 @@ import (
 	"github.com/cockroachdb/cockroach/pkg/ccl/utilccl"
 	"github.com/cockroachdb/cockroach/pkg/sql/parser"
 	"github.com/cockroachdb/cockroach/pkg/testutils/serverutils"
-	"github.com/cockroachdb/cockroach/pkg/testutils/skip"
 	"github.com/cockroachdb/cockroach/pkg/testutils/sqlutils"
 	"github.com/cockroachdb/cockroach/pkg/util/leaktest"
 	"github.com/cockroachdb/cockroach/pkg/util/randutil"
@@ -140,7 +139,6 @@ func TestRandTableInserts(t *testing.T) {
 // sometimes put them into bad states that the parser would never do.
 func TestGenerateParse(t *testing.T) {
 	defer leaktest.AfterTest(t)()
-	skip.WithIssue(t, 66723, "flaky test")
 	defer utilccl.TestingEnableEnterprise()()
 
 	ctx := context.Background()

